clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndTimestamp.java

```diff
@@ -20,14 +20,13 @@
 
 /**
  * A container class for offset and timestamp.
- *
- * Offset must be non-negative.
- *
- * The timestamp should never be negative, unless it is invalid.  This could happen when
- * handling a response from a broker that doesn't support KIP-79.
  */
 public final class OffsetAndTimestamp {
+    // The timestamp should never be negative, unless it is invalid.  This could happen when
```

Comment:
I agree that allowing negatives here feels uncomfortable.  Unfortunately, it reflects the fundamental fact that old servers are not sending back timestamps, and new servers are.  I tried to create a totally separate code path for the old servers, but it was simply too much code duplication.  Also unfortunately, because we need to read the offset from the structures, we cannot use static instances.

Probably the best way around this problem is to stop using the public type internally.  Then we can have an internal type where the timestamp is optional, and a public type where it is not.  This also makes it less likely that we'll break the public API by making an internal change, so it shoudl be a good thing overall.  The downside is that copying will be needed, but I don't think the amount of copying is significant for this use-case.